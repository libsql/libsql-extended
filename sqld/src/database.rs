use std::sync::Arc;

use crate::connection::libsql::LibSqlConnection;
use crate::connection::write_proxy::WriteProxyConnection;
use crate::connection::{Connection, MakeConnection, TrackedConnection};
use crate::replication::ReplicationLogger;

pub struct DatabaseInfo {
    current_frame_no: FrameNo,
}

pub trait Database: Sync + Send + 'static {
    /// The connection type of the database
    type Connection: Connection;

    fn connection_maker(&self) -> Arc<dyn MakeConnection<Connection = Self::Connection>>;
    fn shutdown(&self);
    fn info(&self) -> DatabaseInfo;
}

pub struct ReplicaDatabase {
    pub connection_maker:
        Arc<dyn MakeConnection<Connection = TrackedConnection<WriteProxyConnection>>>,
}

impl Database for ReplicaDatabase {
    type Connection = TrackedConnection<WriteProxyConnection>;

    fn connection_maker(&self) -> Arc<dyn MakeConnection<Connection = Self::Connection>> {
        self.connection_maker.clone()
    }

    fn shutdown(&self) {}

    fn info(&self) -> DatabaseInfo {
        DatabaseInfo { 
            current_frame_no: todo!()
        }
    }
}

pub struct PrimaryDatabase {
    pub logger: Arc<ReplicationLogger>,
    pub connection_maker: Arc<dyn MakeConnection<Connection = TrackedConnection<LibSqlConnection>>>,
}

impl Database for PrimaryDatabase {
    type Connection = TrackedConnection<LibSqlConnection>;

    fn connection_maker(&self) -> Arc<dyn MakeConnection<Connection = Self::Connection>> {
        self.connection_maker.clone()
    }

    fn shutdown(&self) {
        self.logger.closed_signal.send_replace(true);
    }

    fn info(&self) -> DatabaseInfo {
        DatabaseInfo { current_frame_no: todo!() }
    }
}
