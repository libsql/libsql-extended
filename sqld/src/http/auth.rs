use anyhow::{anyhow, Result};
use hyper::{Body, Request};
use std::sync::Arc;

/// HTTP request authorizer.
pub trait Authorizer {
    fn is_authorized(&self, req: &Request<Body>) -> bool;
}

pub fn parse_auth(auth: Option<String>) -> Result<Arc<dyn Authorizer + Sync + Send>> {
    match auth {
        Some(auth) => match auth.split_once(':') {
            Some((scheme, param)) => match scheme {
                "basic" => Ok(Arc::new(BasicAuthAuthorizer {
                    expected_auth: format!("Basic {}", param).to_lowercase(),
                })),
                "bearer" => Ok(Arc::new(BasicAuthAuthorizer {
                    expected_auth: format!("Bearer {}", param).to_lowercase(),
                })),
                _ => Err(anyhow!("unsupported HTTP auth scheme: {}", scheme)),
            },
            None if auth == "always" => Ok(Arc::new(AlwaysAllowAuthorizer {})),
            None => Err(anyhow!("invalid HTTP auth config: {}", auth)),
        },
        None => Ok(Arc::new(AlwaysAllowAuthorizer {})),
    }
}

/// An authorizer that always allows all requests.
pub struct AlwaysAllowAuthorizer {}

impl Authorizer for AlwaysAllowAuthorizer {
    fn is_authorized(&self, _req: &Request<Body>) -> bool {
        true
    }
}

/// Basic authentication authorizer, simply verifying
/// if the header is equal to the expected value.
pub struct BasicAuthAuthorizer {
    // Expected value in `Authorization` header.
    expected_auth: String,
}

impl Authorizer for BasicAuthAuthorizer {
    fn is_authorized(&self, req: &Request<Body>) -> bool {
        let headers = req.headers();
        let actual_auth = headers.get(hyper::header::AUTHORIZATION);
        if let Some(actual_auth) = actual_auth {
            actual_auth
                .to_str()
                .map(|actual_auth| actual_auth.to_lowercase() == self.expected_auth)
                .unwrap_or(false)
        } else {
            false
        }
    }
}
