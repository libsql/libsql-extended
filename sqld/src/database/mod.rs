use std::sync::Arc;

use crate::auth::Authenticated;
use crate::query::{Params, Query, QueryResult};
use crate::query_analysis::{State, Statement};
use crate::Result;

pub mod dump;
pub mod factory;
pub mod libsql;
pub mod write_proxy;

const TXN_TIMEOUT_SECS: u64 = 5;

#[derive(Debug, Clone)]
pub struct Program {
    pub steps: Arc<Vec<Step>>,
}

impl Program {
    pub fn new(steps: Vec<Step>) -> Self {
        Self {
            steps: Arc::new(steps),
        }
    }

    pub fn is_read_only(&self) -> bool {
        self.steps.iter().all(|s| s.query.stmt.is_read_only())
    }

    pub fn steps(&self) -> &[Step] {
        self.steps.as_slice()
    }
}

#[derive(Debug, Clone)]
pub struct Step {
    pub cond: Option<Cond>,
    pub query: Query,
}

#[derive(Debug, Clone)]
pub enum Cond {
    Ok { step: usize },
    Err { step: usize },
    Not { cond: Box<Self> },
    Or { conds: Vec<Self> },
    And { conds: Vec<Self> },
}

pub type DescribeResult = Result<DescribeResponse>;

#[derive(Debug, Clone)]
pub struct DescribeResponse {
    pub params: Vec<DescribeParam>,
    pub cols: Vec<DescribeCol>,
    pub is_explain: bool,
    pub is_readonly: bool,
}

#[derive(Debug, Clone)]
pub struct DescribeParam {
    pub name: Option<String>,
}

#[derive(Debug, Clone)]
pub struct DescribeCol {
    pub name: String,
    pub decltype: Option<String>,
}

#[async_trait::async_trait]
pub trait Database: Send + Sync {
    /// Executes a query program
    async fn execute_program(
        &self,
        pgm: Program,
        auth: Authenticated,
    ) -> Result<(Vec<Option<QueryResult>>, State)>;

    /// Unconditionnaly execute a query as part of a program
    async fn execute_one(&self, query: Query, auth: Authenticated) -> Result<(QueryResult, State)> {
        let pgm = Program::new(vec![Step { cond: None, query }]);

        let (results, state) = self.execute_program(pgm, auth).await?;
        Ok((results.into_iter().next().unwrap().unwrap(), state))
    }

    /// Execute all the queries in the batch sequentially.
    /// If an query in the batch fails, the remaining queries are ignores, and the batch current
    /// transaction (if any) is rolledback.
    async fn execute_batch_or_rollback(
        &self,
        batch: Vec<Query>,
        auth: Authenticated,
    ) -> Result<(Vec<Option<QueryResult>>, State)> {
        let mut steps = make_batch_program(batch);

        if !steps.is_empty() {
            // We add a conditional rollback step if the last step was not sucessful.
            steps.push(Step {
                query: Query {
                    stmt: Statement::parse("ROLLBACK").next().unwrap().unwrap(),
                    params: Params::empty(),
                    want_rows: false,
                },
                cond: Some(Cond::Not {
                    cond: Box::new(Cond::Ok {
                        step: steps.len() - 1,
                    }),
                }),
            })
        }

        let pgm = Program::new(steps);

        let (mut results, state) = self.execute_program(pgm, auth).await?;
        // remove the rollback result
        results.pop();

        Ok((results, state))
    }

    /// Execute all the queries in the batch sequentially.
    /// If an query in the batch fails, the remaining queries are ignored
    async fn execute_batch(
        &self,
        batch: Vec<Query>,
        auth: Authenticated,
    ) -> Result<(Vec<Option<QueryResult>>, State)> {
        let steps = make_batch_program(batch);
        let pgm = Program::new(steps);
        self.execute_program(pgm, auth).await
    }

    async fn rollback(&self, auth: Authenticated) -> Result<()> {
        let (results, _) = self
            .execute_one(
                Query {
                    stmt: Statement::parse("ROLLBACK").next().unwrap().unwrap(),
                    params: Params::empty(),
                    want_rows: false,
                },
                auth,
            )
            .await?;

        results?;

        Ok(())
    }

    /// Parse the SQL statement and return information about it.
    async fn describe(&self, sql: String, auth: Authenticated) -> Result<DescribeResult>;
}

fn make_batch_program(batch: Vec<Query>) -> Vec<Step> {
    let mut steps = Vec::with_capacity(batch.len());
    for (i, query) in batch.into_iter().enumerate() {
        let cond = if i > 0 {
            // only execute if the previous step was a success
            Some(Cond::Ok { step: i - 1 })
        } else {
            None
        };

        let step = Step { cond, query };
        steps.push(step);
    }
    steps
}
