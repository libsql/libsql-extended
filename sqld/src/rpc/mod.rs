use anyhow::Context;
use std::net::SocketAddr;
use std::path::PathBuf;
use std::sync::Arc;
use tower::util::option_layer;

use crate::namespace::{NamespaceStore, PrimaryNamespaceMaker};
use crate::rpc::proxy::rpc::proxy_server::ProxyServer;
use crate::rpc::proxy::ProxyService;
pub use crate::rpc::replication_log::rpc::replication_log_server::ReplicationLogServer;
use crate::rpc::replication_log::ReplicationLogService;
use crate::utils::services::idle_shutdown::IdleShutdownLayer;

pub mod proxy;
pub mod replication_log;
pub mod replication_log_proxy;

#[allow(clippy::too_many_arguments)]
pub async fn run_rpc_server(
    addr: SocketAddr,
    tls: bool,
    cert_path: Option<PathBuf>,
    key_path: Option<PathBuf>,
    ca_cert_path: Option<PathBuf>,
    idle_shutdown_layer: Option<IdleShutdownLayer>,
    namespaces: Arc<NamespaceStore<PrimaryNamespaceMaker>>,
    disable_namespaces: bool,
) -> anyhow::Result<()> {
    let proxy_service = ProxyService::new(namespaces.clone());
    let logger_service = ReplicationLogService::new(
        namespaces.clone(),
        idle_shutdown_layer.clone(),
        None,
        disable_namespaces,
    );

    tracing::info!("serving write proxy server at {addr}");

    let mut builder = tonic::transport::Server::builder();
    if tls {
        let cert_pem = std::fs::read_to_string(cert_path.unwrap())?;
        let key_pem = std::fs::read_to_string(key_path.unwrap())?;
        let identity = tonic::transport::Identity::from_pem(cert_pem, key_pem);

        let ca_cert_pem = std::fs::read_to_string(ca_cert_path.unwrap())?;
        let ca_cert = tonic::transport::Certificate::from_pem(ca_cert_pem);

        let tls_config = tonic::transport::ServerTlsConfig::new()
            .identity(identity)
            .client_ca_root(ca_cert);
        builder = builder
            .tls_config(tls_config)
            .context("Failed to read the TSL config of RPC server")?;
    }
    builder
        .layer(&option_layer(idle_shutdown_layer))
        .add_service(ProxyServer::new(proxy_service))
        .add_service(ReplicationLogServer::new(logger_service))
        .serve(addr)
        .await?;

    Ok(())
}
