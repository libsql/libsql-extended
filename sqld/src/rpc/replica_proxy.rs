use std::sync::Arc;

use hyper::Uri;
use tonic::{transport::Channel, Request, Status};

use crate::auth::Auth;

use super::proxy::rpc::{
    self, proxy_client::ProxyClient, proxy_server::Proxy, Ack, DescribeRequest, DescribeResult,
    DisconnectMessage, ExecuteResults,
};

pub struct ReplicaProxyService {
    client: ProxyClient<Channel>,
    auth: Arc<Auth>,
}

impl ReplicaProxyService {
    pub fn new(channel: Channel, uri: Uri, auth: Arc<Auth>) -> Self {
        let client = ProxyClient::with_origin(channel, uri);
        Self { client, auth }
    }

    fn do_auth<T>(&self, req: &mut Request<T>) -> Result<(), Status> {
        let authenticated = self.auth.authenticate_grpc(req, false)?;

        authenticated.upgrade_grpc_request(req);

        Ok(())
    }
}

#[tonic::async_trait]
impl Proxy for ReplicaProxyService {
    async fn execute(
        &self,
        mut req: tonic::Request<rpc::ProgramReq>,
    ) -> Result<tonic::Response<ExecuteResults>, tonic::Status> {
        self.do_auth(&mut req)?;

        let mut client = self.client.clone();
        client.execute(req).await
    }

    //TODO: also handle cleanup on peer disconnect
    async fn disconnect(
        &self,
        mut msg: tonic::Request<DisconnectMessage>,
    ) -> Result<tonic::Response<Ack>, tonic::Status> {
        self.do_auth(&mut msg)?;

        let mut client = self.client.clone();
        client.disconnect(msg).await
    }

    async fn describe(
        &self,
        mut req: tonic::Request<DescribeRequest>,
    ) -> Result<tonic::Response<DescribeResult>, tonic::Status> {
        self.do_auth(&mut req)?;

        let mut client = self.client.clone();
        client.describe(req).await
    }
}
