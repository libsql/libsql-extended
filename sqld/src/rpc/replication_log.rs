pub mod rpc {
    #![allow(clippy::all)]
    tonic::include_proto!("wal_log");
}

use std::collections::HashSet;
use std::net::SocketAddr;
use std::pin::Pin;
use std::sync::{Arc, RwLock};

use futures::stream::BoxStream;
use tokio::sync::mpsc;
use tokio_stream::wrappers::ReceiverStream;
use tokio_stream::StreamExt;
use tonic::Status;

use crate::auth::Auth;
use crate::namespace::{NamespaceName, NamespaceStore, PrimaryNamespaceMaker};
use crate::replication::primary::frame_stream::FrameStream;
use crate::replication::LogReadError;
use crate::utils::services::idle_shutdown::IdleShutdownKicker;
use crate::BLOCKING_RT;

use self::rpc::replication_log_server::ReplicationLog;
use self::rpc::{Frame, Frames, HelloRequest, HelloResponse, LogOffset};

use super::NAMESPACE_DOESNT_EXIST;

pub struct ReplicationLogService {
    namespaces: NamespaceStore<PrimaryNamespaceMaker>,
    replicas_with_hello: RwLock<HashSet<(SocketAddr, NamespaceName)>>,
    idle_shutdown_layer: Option<IdleShutdownKicker>,
    auth: Option<Arc<Auth>>,
    disable_namespaces: bool,
}

pub const NO_HELLO_ERROR_MSG: &str = "NO_HELLO";
pub const NEED_SNAPSHOT_ERROR_MSG: &str = "NEED_SNAPSHOT";

pub const MAX_FRAMES_PER_BATCH: usize = 1024;

impl ReplicationLogService {
    pub fn new(
        namespaces: NamespaceStore<PrimaryNamespaceMaker>,
        idle_shutdown_layer: Option<IdleShutdownKicker>,
        auth: Option<Arc<Auth>>,
        disable_namespaces: bool,
    ) -> Self {
        Self {
            namespaces,
            replicas_with_hello: Default::default(),
            idle_shutdown_layer,
            auth,
            disable_namespaces,
        }
    }

    fn authenticate<T>(&self, req: &tonic::Request<T>) -> Result<(), Status> {
        if let Some(auth) = &self.auth {
            let _ = auth.authenticate_grpc(req, self.disable_namespaces)?;
        }

        Ok(())
    }
}

fn map_frame_stream_output(
    r: Result<crate::replication::frame::Frame, LogReadError>,
) -> Result<Frame, Status> {
    match r {
        Ok(frame) => Ok(Frame {
            data: frame.bytes(),
        }),
        Err(LogReadError::SnapshotRequired) => Err(Status::new(
            tonic::Code::FailedPrecondition,
            NEED_SNAPSHOT_ERROR_MSG,
        )),
        Err(LogReadError::Error(e)) => Err(Status::new(tonic::Code::Internal, e.to_string())),
        // this error should be caught before, but we handle it nicely anyways
        Err(LogReadError::Ahead) => Err(Status::new(
            tonic::Code::OutOfRange,
            "frame not yet available",
        )),
    }
}

pub struct StreamGuard<S> {
    s: S,
    idle_shutdown_layer: Option<IdleShutdownKicker>,
}

impl<S> StreamGuard<S> {
    fn new(s: S, mut idle_shutdown_layer: Option<IdleShutdownKicker>) -> Self {
        if let Some(isl) = idle_shutdown_layer.as_mut() {
            isl.add_connected_replica()
        }
        Self {
            s,
            idle_shutdown_layer,
        }
    }
}

impl<S> Drop for StreamGuard<S> {
    fn drop(&mut self) {
        if let Some(isl) = self.idle_shutdown_layer.as_mut() {
            isl.remove_connected_replica()
        }
    }
}

impl<S: futures::stream::Stream + Unpin> futures::stream::Stream for StreamGuard<S> {
    type Item = S::Item;

    fn poll_next(
        self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Option<Self::Item>> {
        Pin::new(&mut self.get_mut().s).poll_next(cx)
    }
}

#[tonic::async_trait]
impl ReplicationLog for ReplicationLogService {
    type LogEntriesStream = BoxStream<'static, Result<Frame, Status>>;
    type SnapshotStream = BoxStream<'static, Result<Frame, Status>>;

    async fn log_entries(
        &self,
        req: tonic::Request<LogOffset>,
    ) -> Result<tonic::Response<Self::LogEntriesStream>, Status> {
        self.authenticate(&req)?;
        let namespace = super::extract_namespace(self.disable_namespaces, &req)?;

        let replica_addr = req
            .remote_addr()
            .ok_or(Status::internal("No remote RPC address"))?;
        let req = req.into_inner();
        {
            let guard = self.replicas_with_hello.read().unwrap();
            if !guard.contains(&(replica_addr, namespace.clone())) {
                return Err(Status::failed_precondition(NO_HELLO_ERROR_MSG));
            }
        }

        let logger = self
            .namespaces
            .with(namespace, |ns| ns.db.logger.clone())
            .await
            .map_err(|e| {
                if let crate::error::Error::NamespaceDoesntExist(_) = e {
                    Status::failed_precondition(NAMESPACE_DOESNT_EXIST)
                } else {
                    Status::internal(e.to_string())
                }
            })?;

        let stream = StreamGuard::new(
            FrameStream::new(logger, req.next_offset, true, None)
                .map_err(|e| Status::internal(e.to_string()))?,
            self.idle_shutdown_layer.clone(),
        )
        .map(map_frame_stream_output);

        Ok(tonic::Response::new(Box::pin(stream)))
    }

    async fn batch_log_entries(
        &self,
        req: tonic::Request<LogOffset>,
    ) -> Result<tonic::Response<Frames>, Status> {
        self.authenticate(&req)?;
        let namespace = super::extract_namespace(self.disable_namespaces, &req)?;

        let replica_addr = req
            .remote_addr()
            .ok_or(Status::internal("No remote RPC address"))?;
        let req = req.into_inner();
        {
            let guard = self.replicas_with_hello.read().unwrap();
            if !guard.contains(&(replica_addr, namespace.clone())) {
                return Err(Status::failed_precondition(NO_HELLO_ERROR_MSG));
            }
        }

        let logger = self
            .namespaces
            .with(namespace, |ns| ns.db.logger.clone())
            .await
            .map_err(|e| {
                if let crate::error::Error::NamespaceDoesntExist(_) = e {
                    Status::failed_precondition(NAMESPACE_DOESNT_EXIST)
                } else {
                    Status::internal(e.to_string())
                }
            })?;

        let frames = StreamGuard::new(
            FrameStream::new(logger, req.next_offset, false, Some(MAX_FRAMES_PER_BATCH))
                .map_err(|e| Status::internal(e.to_string()))?,
            self.idle_shutdown_layer.clone(),
        )
        .map(map_frame_stream_output)
        .collect::<Result<Vec<_>, _>>()
        .await?;

        Ok(tonic::Response::new(Frames { frames }))
    }

    async fn hello(
        &self,
        req: tonic::Request<HelloRequest>,
    ) -> Result<tonic::Response<HelloResponse>, Status> {
        self.authenticate(&req)?;
        let namespace = super::extract_namespace(self.disable_namespaces, &req)?;

        use tonic::transport::server::TcpConnectInfo;

        req.extensions().get::<TcpConnectInfo>().unwrap();
        let replica_addr = req
            .remote_addr()
            .ok_or(Status::internal("No remote RPC address"))?;

        {
            let mut guard = self.replicas_with_hello.write().unwrap();
            guard.insert((replica_addr, namespace.clone()));
        }

        let logger = self
            .namespaces
            .with(namespace, |ns| ns.db.logger.clone())
            .await
            .map_err(|e| {
                if let crate::error::Error::NamespaceDoesntExist(_) = e {
                    Status::failed_precondition(NAMESPACE_DOESNT_EXIST)
                } else {
                    Status::internal(e.to_string())
                }
            })?;

        let response = HelloResponse {
            log_id: logger.log_id().to_string(),
        };

        Ok(tonic::Response::new(response))
    }

    async fn snapshot(
        &self,
        req: tonic::Request<LogOffset>,
    ) -> Result<tonic::Response<Self::SnapshotStream>, Status> {
        self.authenticate(&req)?;
        let namespace = super::extract_namespace(self.disable_namespaces, &req)?;

        let (sender, receiver) = mpsc::channel(10);
        let req = req.into_inner();
        let logger = self
            .namespaces
            .with(namespace, |ns| ns.db.logger.clone())
            .await
            .unwrap();
        let offset = req.next_offset;
        match BLOCKING_RT
            .spawn_blocking(move || logger.get_snapshot_file(offset))
            .await
        {
            Ok(Ok(Some(snapshot))) => {
                BLOCKING_RT.spawn_blocking(move || {
                    let size_after = snapshot.header().size_after;
                    let mut frames = snapshot.frames_iter_from(offset).peekable();
                    loop {
                        match frames.next() {
                            Some(Ok(mut frame)) => {
                                // this is the last frame we're sending for this snapshot, set the
                                // frame_no
                                if frames.peek().is_none() {
                                    frame.header_mut().size_after = size_after;
                                }
                                let _ = sender.blocking_send(Ok(Frame {
                                    data: crate::replication::frame::Frame::from(frame).bytes(),
                                }));
                            }
                            Some(Err(e)) => {
                                let _ = sender.blocking_send(Err(Status::new(
                                    tonic::Code::Internal,
                                    e.to_string(),
                                )));
                                break;
                            }
                            None => {
                                break;
                            }
                        }
                    }
                });

                Ok(tonic::Response::new(Box::pin(ReceiverStream::new(
                    receiver,
                ))))
            }
            Ok(Ok(None)) => Err(Status::new(tonic::Code::Unavailable, "snapshot not found")),
            Err(e) => Err(Status::new(tonic::Code::Internal, e.to_string())),
            Ok(Err(e)) => Err(Status::new(tonic::Code::Internal, e.to_string())),
        }
    }
}
