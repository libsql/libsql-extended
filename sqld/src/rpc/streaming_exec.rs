use std::sync::Arc;

use futures_core::Stream;
use futures_option::OptionExt;
use prost::Message;
use rusqlite::types::ValueRef;
use tokio::pin;
use tokio::sync::mpsc;
use tokio_stream::StreamExt;
use tonic::{Code, Status};

use crate::auth::Authenticated;
use crate::connection::Connection;
use crate::database::PrimaryConnection;
use crate::query_analysis::TxnStatus;
use crate::query_result_builder::{
    Column, QueryBuilderConfig, QueryResultBuilder, QueryResultBuilderError,
};
use crate::replication::FrameNo;
use crate::rpc::proxy::rpc::exec_req::Request;
use crate::rpc::proxy::rpc::exec_resp::{self, Response};

use super::proxy::rpc::resp_step::Step;
use super::proxy::rpc::{self, ExecReq, ExecResp, ProgramResp, RespStep, RowValue};

pub fn make_proxy_stream<S>(conn: PrimaryConnection, auth: Authenticated, request_stream: S) -> impl Stream<Item = Result<ExecResp, Status>> 
where
    S: Stream<Item = Result<ExecReq, Status>>,
{
    async_stream::stream! {
        let mut current_request_fut = None;
        let (snd, mut recv) = mpsc::channel(1);
        let conn = Arc::new(conn);
        pin!(request_stream);

        loop {
            tokio::select! {
                biased;
                Some(maybe_req) = request_stream.next() => {
                    match maybe_req {
                        Err(e) => {
                            tracing::error!("stream error: {e}");
                            break
                        }
                        Ok(req) => {
                            let request_id = req.request_id;
                            match req.request {
                                Some(Request::Execute(pgm)) => {
                                    let Ok(pgm) =
                                        crate::connection::program::Program::try_from(pgm.pgm.unwrap()) else {
                                            yield Err(Status::new(Code::InvalidArgument, "invalid program"));
                                            break
                                        };
                                    let conn = conn.clone();
                                    let auth = auth.clone();
                                    let sender = snd.clone();

                                    let fut = async move {
                                        let builder = StreamResponseBuilder {
                                            request_id,
                                            sender,
                                            current: None,
                                            current_size: 0,
                                        };

                                        let ret = conn.execute_program(pgm, auth, builder, None).await;
                                        (ret, request_id)
                                    };

                                    current_request_fut.replace(Box::pin(fut));
                                }
                                Some(Request::Describe(_)) => todo!(),
                                None => {
                                    yield Err(Status::new(Code::InvalidArgument, "invalid request"));
                                    break
                                }
                            }
                        }
                    }
                },
                Some(res) = recv.recv() => {
                    yield Ok(res);
                },
                (ret, request_id) = current_request_fut.current(), if current_request_fut.is_some() => {
                    if let Err(e) = ret {
                        yield Ok(ExecResp { request_id, response: Some(Response::Error(e.into())) })
                    }
                },
                else => break,
            }
        }
    }
}

struct StreamResponseBuilder {
    request_id: u32,
    sender: mpsc::Sender<ExecResp>,
    current: Option<ProgramResp>,
    current_size: usize,
}

impl StreamResponseBuilder {
    fn current(&mut self) -> &mut ProgramResp {
        self.current
            .get_or_insert_with(|| ProgramResp { steps: Vec::new() })
    }

    fn push(&mut self, step: Step) -> Result<(), QueryResultBuilderError> {
        const MAX_RESPONSE_SIZE: usize = bytesize::ByteSize::mb(1).as_u64() as usize;

        let current = self.current();
        let step = RespStep { step: Some(step) };
        let size = step.encoded_len();
        current.steps.push(step);
        self.current_size += size;

        if self.current_size >= MAX_RESPONSE_SIZE {
            self.flush()?;
        }

        Ok(())
    }

    fn flush(&mut self) -> Result<(), QueryResultBuilderError> {
        if let Some(current) = self.current.take() {
            let resp = ExecResp {
                request_id: self.request_id,
                response: Some(exec_resp::Response::ProgramResp(current)),
            };
            self.current_size = 0;
            self.sender
                .blocking_send(resp)
                .map_err(|_| QueryResultBuilderError::Internal(anyhow::anyhow!("stream closed")))?;
        }

        Ok(())
    }
}

impl QueryResultBuilder for StreamResponseBuilder {
    type Ret = ();

    fn init(&mut self, _config: &QueryBuilderConfig) -> Result<(), QueryResultBuilderError> {
        self.push(Step::Init(rpc::Init {}))?;
        Ok(())
    }

    fn begin_step(&mut self) -> Result<(), QueryResultBuilderError> {
        self.push(Step::BeginStep(rpc::BeginStep {}))?;
        Ok(())
    }

    fn finish_step(
        &mut self,
        affected_row_count: u64,
        last_insert_rowid: Option<i64>,
    ) -> Result<(), QueryResultBuilderError> {
        self.push(Step::FinishStep(rpc::FinishStep {
            affected_row_count,
            last_insert_rowid,
        }))?;
        Ok(())
    }

    fn step_error(&mut self, error: crate::error::Error) -> Result<(), QueryResultBuilderError> {
        self.push(Step::StepError(rpc::StepError {
            error: Some(error.into()),
        }))?;
        Ok(())
    }

    fn cols_description<'a>(
        &mut self,
        cols: impl IntoIterator<Item = impl Into<Column<'a>>>,
    ) -> Result<(), QueryResultBuilderError> {
        self.push(Step::ColsDescription(rpc::ColsDescription {
            columns: cols
                .into_iter()
                .map(Into::into)
                .map(|c| rpc::Column {
                    name: c.name.into(),
                    decltype: c.decl_ty.map(Into::into),
                })
                .collect::<Vec<_>>(),
        }))?;
        Ok(())
    }

    fn begin_rows(&mut self) -> Result<(), QueryResultBuilderError> {
        self.push(Step::BeginRows(rpc::BeginRows {}))?;
        Ok(())
    }

    fn begin_row(&mut self) -> Result<(), QueryResultBuilderError> {
        self.push(Step::BeginRow(rpc::BeginRow {}))?;
        Ok(())
    }

    fn add_row_value(&mut self, v: ValueRef) -> Result<(), QueryResultBuilderError> {
        self.push(Step::AddRowValue(rpc::AddRowValue {
            val: Some(v.into()),
        }))?;
        Ok(())
    }

    fn finish_row(&mut self) -> Result<(), QueryResultBuilderError> {
        self.push(Step::FinishRow(rpc::FinishRow {}))?;
        Ok(())
    }

    fn finish_rows(&mut self) -> Result<(), QueryResultBuilderError> {
        self.push(Step::FinishRows(rpc::FinishRows {}))?;
        Ok(())
    }

    fn finish(
        &mut self,
        last_frame_no: Option<FrameNo>,
        state: TxnStatus,
    ) -> Result<(), QueryResultBuilderError> {
        self.push(Step::Finish(rpc::Finish {
            last_frame_no,
            state: rpc::State::from(state).into(),
        }))?;
        self.flush()?;
        Ok(())
    }

    fn into_ret(self) -> Self::Ret { }
}

impl From<ValueRef<'_>> for RowValue {
    fn from(value: ValueRef<'_>) -> Self {
        use rpc::row_value::Value;

        let value = Some(match value {
            ValueRef::Null => Value::Null(true),
            ValueRef::Integer(i) => Value::Integer(i),
            ValueRef::Real(x) => Value::Real(x),
            ValueRef::Text(s) => Value::Text(String::from_utf8(s.to_vec()).unwrap()),
            ValueRef::Blob(b) => Value::Blob(b.to_vec()),
        });

        RowValue { value }
    }
}
