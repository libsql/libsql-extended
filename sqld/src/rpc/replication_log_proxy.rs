use hyper::Uri;
use tonic::{transport::Channel, Status};

use super::replication_log::rpc::replication_log_client::ReplicationLogClient;
use super::replication_log::rpc::replication_log_server::ReplicationLog;
use super::replication_log::rpc::{Frame, Frames, HelloRequest, HelloResponse, LogOffset};

/// A replication log service that proxies request to the primary.
pub struct ReplicationLogProxyService {
    client: ReplicationLogClient<Channel>,
}

impl ReplicationLogProxyService {
    pub fn new(channel: Channel, uri: Uri) -> Self {
        let client =
            ReplicationLogClient::with_origin(channel, uri).max_decoding_message_size(usize::MAX);

        Self { client }
    }
}

#[tonic::async_trait]
impl ReplicationLog for ReplicationLogProxyService {
    type LogEntriesStream = tonic::codec::Streaming<Frame>;
    type SnapshotStream = tonic::codec::Streaming<Frame>;

    async fn log_entries(
        &self,
        req: tonic::Request<LogOffset>,
    ) -> Result<tonic::Response<Self::LogEntriesStream>, Status> {
        let mut client = self.client.clone();
        client.log_entries(req).await
    }

    async fn batch_log_entries(
        &self,
        req: tonic::Request<LogOffset>,
    ) -> Result<tonic::Response<Frames>, Status> {
        let mut client = self.client.clone();
        client.batch_log_entries(req).await
    }

    async fn hello(
        &self,
        req: tonic::Request<HelloRequest>,
    ) -> Result<tonic::Response<HelloResponse>, Status> {
        let mut client = self.client.clone();
        client.hello(req).await
    }

    async fn snapshot(
        &self,
        req: tonic::Request<LogOffset>,
    ) -> Result<tonic::Response<Self::SnapshotStream>, Status> {
        let mut client = self.client.clone();
        client.snapshot(req).await
    }
}
