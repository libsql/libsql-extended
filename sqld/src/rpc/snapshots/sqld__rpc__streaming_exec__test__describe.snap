---
source: sqld/src/rpc/streaming_exec.rs
expression: stream.next().await.unwrap().unwrap()
---
ExecResp {
    request_id: 0,
    response: Some(
        DescribeResp(
            DescribeResp {
                params: [
                    DescribeParam {
                        name: Some(
                            "$hello",
                        ),
                    },
                ],
                cols: [
                    DescribeCol {
                        name: "$hello",
                        decltype: None,
                    },
                ],
                is_explain: false,
                is_readonly: true,
            },
        ),
    ),
}
